src/tools/immutable_object_cache/ObjectCacheStore.cc

```diff
@@ -0,0 +1,244 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "ObjectCacheStore.h"
+#include "Utils.h"
+#include <experimental/filesystem>
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_immutable_obj_cache
+#undef dout_prefix
+#define dout_prefix *_dout << "ceph::cache::ObjectCacheStore: " << this << " " \
+                           << __func__ << ": "
+
+namespace efs = std::experimental::filesystem;
+
+namespace ceph {
+namespace immutable_obj_cache {
+
+ObjectCacheStore::ObjectCacheStore(CephContext *cct, ContextWQ* work_queue)
+      : m_cct(cct), m_work_queue(work_queue), m_rados(new librados::Rados()),
+        m_ioctxs_lock("ceph::cache::ObjectCacheStore::m_ioctxs_lock") {
+
+  object_cache_max_size =
+    m_cct->_conf.get_val<Option::size_t>("immutable_object_cache_max_size");
+
+  std::string cache_path = m_cct->_conf.get_val<std::string>("immutable_object_cache_path");
+  m_cache_root_dir = cache_path + "/ceph_immutable_obj_cache/";
+
+  //TODO(): allow to set cache level
+  m_policy = new SimplePolicy(m_cct, object_cache_max_size/(4096*1024), 0.1);
+}
+
+ObjectCacheStore::~ObjectCacheStore() {
+  delete m_policy;
+}
+
+int ObjectCacheStore::init(bool reset) {
+  ldout(m_cct, 20) << dendl;
+
+  int ret = m_rados->init_with_context(m_cct);
+  if(ret < 0) {
+    lderr(m_cct) << "fail to init Ceph context" << dendl;
+    return ret;
+  }
+
+  ret = m_rados->connect();
+  if(ret < 0 ) {
+    lderr(m_cct) << "fail to conect to cluster" << dendl;
+    return ret;
+  }
+
+  //TODO(): fsck and reuse existing cache objects
+  if (reset) {
+    if (efs::exists(m_cache_root_dir)) {
+      efs::remove_all(m_cache_root_dir);
+    }
+    efs::create_directories(m_cache_root_dir);
```

Comment:
Probably this can be improved later, anyway, leaving a comment here.

I am not sure it is a good idea to make parent directories if they do not exist -- the daemon may not have enough rights for this. Removing the root directory may be not a very good idea too, because it may be a mount point, or again the daemon may not have rights for this.

Probably it is safer only to clean the layout sub directories without trying to remove the root directory?

And it looks like it may throw exception on an error?

Reply:
Thanks for the inputs! Currently I did add a sub dir "/ceph_immutable_obj_cache/" under the cache dir, so it will only remove this sub dir. 
https://github.com/ceph/ceph/blob/c4d1fd0594d721853953fc7e196e5310a82d6efd/src/tools/immutable_object_cache/ObjectCacheStore.cc#L27

Should the daemon do some pre-check, if we got denied on read/write, then we should refuse to start? right, should also try/catch any exception here.